File: backend/app/services/auth_service.py
from __future__ import annotations

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.core.security import (
    create_access_token,
    create_refresh_token,
    decode_token,
    hash_password,
    verify_password,
)
from app.models.settings import Settings
from app.models.user import User
from app.repositories.users import UserRepository
from app.schemas.auth import PasswordResetConfirm, UserLogin, UserRegister
from app.schemas.user import UserAdminCreate, UserAdminUpdate
from app.services.workspace_service import WorkspaceService


class AuthService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.users = UserRepository(db)

    def register(self, payload: UserRegister) -> User:
        if self.users.get_by_email(payload.email):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Email already registered")

        is_first_user = self.users.count() == 0
        workspace = WorkspaceService(self.db).get_or_create()
        if not is_first_user and not workspace.public_registration_enabled:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Public registration is disabled. Ask an administrator to create your account.",
            )

        user = User(
            email=payload.email,
            full_name=payload.full_name,
            hashed_password=hash_password(payload.password),
            is_superuser=is_first_user,
        )
        settings = Settings(user=user)
        self.db.add_all([user, settings])
        self.db.commit()
        self.db.refresh(user)
        return user

    def create_user(self, actor: User, payload: UserAdminCreate) -> User:
        if not actor.is_superuser:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Admin access required")
        if self.users.get_by_email(payload.email):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Email already registered")

        user = User(
            email=payload.email,
            full_name=payload.full_name,
            hashed_password=hash_password(payload.password),
            is_superuser=payload.is_superuser,
            is_active=payload.is_active,
        )
        settings = Settings(user=user)
        self.db.add_all([user, settings])
        self.db.commit()
        self.db.refresh(user)
        return user

    def update_user(self, actor: User, user_id: str, payload: UserAdminUpdate) -> User:
        if not actor.is_superuser:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Admin access required")

        user = self.users.get(user_id)
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

        updates = payload.model_dump(exclude_unset=True)

        next_email = updates.get("email")
        if next_email and next_email != user.email and self.users.get_by_email(next_email):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Email already registered")

        next_is_superuser = updates.get("is_superuser", user.is_superuser)
        next_is_active = updates.get("is_active", user.is_active)
        if user.is_superuser and not next_is_superuser and self.users.count_superusers() <= 1:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="The last admin cannot be changed to a non-admin user.",
            )
        if user.is_superuser and not next_is_active and self.users.count_superusers() <= 1:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="The last admin cannot be made inactive.",
            )

        if "email" in updates:
            user.email = updates["email"]
        if "full_name" in updates:
            user.full_name = updates["full_name"]
        if "password" in updates and updates["password"]:
            user.hashed_password = hash_password(updates["password"])
        if "is_superuser" in updates:
            user.is_superuser = updates["is_superuser"]
        if "is_active" in updates:
            user.is_active = updates["is_active"]

        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)
        return user

    def login(self, payload: UserLogin) -> tuple[User, str, str]:
        workspace = WorkspaceService(self.db).get_or_create()
        if not workspace.authentication_enabled:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Authentication is disabled for this workspace.",
            )
        user = self.users.get_by_email(payload.email)
        if not user or not verify_password(payload.password, user.hashed_password):
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid credentials")
        return user, create_access_token(user.id), create_refresh_token(user.id)

    def refresh(self, refresh_token: str) -> str:
        workspace = WorkspaceService(self.db).get_or_create()
        if not workspace.authentication_enabled:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Authentication is disabled for this workspace.",
            )
        try:
            payload = decode_token(refresh_token, refresh=True)
        except Exception as exc:  # noqa: BLE001
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid refresh token") from exc
        return create_access_token(payload["sub"])

    def request_password_reset(self, email: str) -> dict[str, str]:
        user = self.users.get_by_email(email)
        if not user:
            return {"message": "If the account exists, a reset token has been generated."}
        token = create_access_token(user.id)
        return {"message": "Reset token generated for development use.", "reset_token": token}

    def confirm_password_reset(self, payload: PasswordResetConfirm) -> dict[str, str]:
        try:
            token_payload = decode_token(payload.token)
        except Exception as exc:  # noqa: BLE001
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid reset token") from exc

        user = self.users.get(token_payload["sub"])
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

        user.hashed_password = hash_password(payload.new_password)
        self.db.add(user)
        self.db.commit()
        return {"message": "Password updated successfully"}

File: core/auth/authService.ts
import { randomUUID } from "node:crypto";

import type { ChikaimaDatabase } from "../db/client.js";
import { settingsTable } from "../db/schema.js";
import { badRequest, forbidden, unauthorized, notFound, serviceUnavailable } from "../errors.js";
import { getConfig } from "../config/index.js";
import { createAccessToken, createPasswordResetToken, createRefreshToken, decodeToken, hashPassword, verifyPassword } from "../security/index.js";
import { UserRepository, type UserRow } from "./repository.js";
import { WorkspaceService } from "./workspaceService.js";

export interface UserRegisterInput {
  email: string;
  fullName: string;
  password: string;
}

export interface UserAdminCreateInput extends UserRegisterInput {
  isSuperuser: boolean;
  isActive: boolean;
}

export interface UserAdminUpdateInput {
  email?: string;
  fullName?: string;
  password?: string;
  isSuperuser?: boolean;
  isActive?: boolean;
}

export interface UserLoginInput {
  email: string;
  password: string;
}

export class AuthService {
  readonly users: UserRepository;

  constructor(private readonly db: ChikaimaDatabase) {
    this.users = new UserRepository(db);
  }

  register(payload: UserRegisterInput): UserRow {
    if (this.users.getByEmail(payload.email)) {
      throw badRequest("Email already registered");
    }

    const isFirstUser = this.users.count() === 0;
    const workspace = new WorkspaceService(this.db).getOrCreate();
    if (!isFirstUser && !workspace.publicRegistrationEnabled) {
      throw forbidden("Public registration is disabled. Ask an administrator to create your account.");
    }

    return this.createUserRow(payload.email, payload.fullName, payload.password, isFirstUser, true);
  }

  createUser(actor: UserRow, payload: UserAdminCreateInput): UserRow {
    if (!actor.isSuperuser) {
      throw forbidden("Admin access required");
    }
    if (this.users.getByEmail(payload.email)) {
      throw badRequest("Email already registered");
    }
    return this.createUserRow(payload.email, payload.fullName, payload.password, payload.isSuperuser, payload.isActive);
  }

  updateUser(actor: UserRow, userId: string, payload: UserAdminUpdateInput): UserRow {
    if (!actor.isSuperuser) {
      throw forbidden("Admin access required");
    }

    const user = this.users.get(userId);
    if (!user) {
      throw notFound("User not found");
    }

    if (payload.email && payload.email !== user.email && this.users.getByEmail(payload.email)) {
      throw badRequest("Email already registered");
    }

    const nextIsSuperuser = payload.isSuperuser ?? user.isSuperuser;
    const nextIsActive = payload.isActive ?? user.isActive;
    if (user.isSuperuser && !nextIsSuperuser && this.users.countSuperusers() <= 1) {
      throw badRequest("The last admin cannot be changed to a non-admin user.");
    }
    if (user.isSuperuser && !nextIsActive && this.users.countSuperusers() <= 1) {
      throw badRequest("The last admin cannot be made inactive.");
    }

    const patch: Partial<Omit<UserRow, "id">> = {};
    if (payload.email !== undefined) patch.email = payload.email;
    if (payload.fullName !== undefined) patch.fullName = payload.fullName;
    if (payload.password) patch.hashedPassword = hashPassword(payload.password);
    if (payload.isSuperuser !== undefined) patch.isSuperuser = payload.isSuperuser;
    if (payload.isActive !== undefined) patch.isActive = payload.isActive;

    return this.users.update(userId, patch);
  }

  async login(payload: UserLoginInput): Promise<{ user: UserRow; accessToken: string; refreshToken: string }> {
    const workspace = new WorkspaceService(this.db).getOrCreate();
    if (!workspace.authenticationEnabled) {
      throw forbidden("Authentication is disabled for this workspace.");
    }

    const user = this.users.getByEmail(payload.email);
    if (user && !user.isActive) {
      throw forbidden("Your account is inactive");
    }
    if (!user || !verifyPassword(payload.password, user.hashedPassword)) {
      throw unauthorized("Invalid credentials");
    }

    const [accessToken, refreshToken] = await Promise.all([createAccessToken(user.id), createRefreshToken(user.id)]);
    return { user, accessToken, refreshToken };
  }

  async refresh(refreshToken: string): Promise<string> {
    const workspace = new WorkspaceService(this.db).getOrCreate();
    if (!workspace.authenticationEnabled) {
      throw forbidden("Authentication is disabled for this workspace.");
    }

    let payload;
    try {
      payload = await decodeToken(refreshToken, { refresh: true, expectedType: "refresh" });
    } catch {
      throw unauthorized("Invalid refresh token");
    }

    return createAccessToken(payload.sub);
  }

  async requestPasswordReset(email: string): Promise<{ message: string }> {
    const config = getConfig();
    if (config.isProduction) {
      throw serviceUnavailable("Password reset is not configured for this deployment.");
    }

    const response = { message: "If the account exists, password reset instructions have been sent." };
    const user = this.users.getByEmail(email);
    if (!user) return response;

    const token = await createPasswordResetToken(user.id);
    console.info(`Password reset token for ${user.email}: ${token}`);

    return response;
  }

  async confirmPasswordReset(payload: { token: string; newPassword: string }): Promise<{ message: string }> {
    let decoded;
    try {
      decoded = await decodeToken(payload.token, { expectedType: "password_reset" });
    } catch {
      throw unauthorized("Invalid reset token");
    }

    const user = this.users.get(decoded.sub);
    if (!user) {
      throw notFound("User not found");
    }

    this.users.update(user.id, { hashedPassword: hashPassword(payload.newPassword) });
    return { message: "Password updated successfully" };
  }

  private createUserRow(email: string, fullName: string, password: string, isSuperuser: boolean, isActive: boolean): UserRow {
    const now = new Date().toISOString();
    const user: UserRow = {
      id: randomUUID(),
      email,
      fullName,
      hashedPassword: hashPassword(password),
      isSuperuser,
      isActive,
      createdAt: now,
      updatedAt: now,
    };
    this.users.insert(user);
    this.db.insert(settingsTable).values({
      id: randomUUID(),
      userId: user.id,
      theme: "dark",
      defaultModelId: null,
      preferences: {},
      createdAt: now,
      updatedAt: now,
    }).run();
    return user;
  }
}
